packages/nx/src/plugins/js/project-graph/build-dependencies/target-project-locator.ts

```diff
@@ -302,6 +316,52 @@ export class TargetProjectLocator {
     const project = findProjectForPath(file, this.projectRootMappings);
     return this.nodes[project];
   }
+
+  /**
+   * In many cases the package.json will be directly resolvable, so we try that first.
+   * If, however, package exports are used and the package.json is not defined, we will
+   * need to resolve the main entry point of the package and traverse upwards to find the
+   * package.json.
+   *
+   * In some cases, such as when multiple module formats are published, the resolved package.json
+   * might only contain the "type" field - no "name" or "version", so in such cases we keep traversing
+   * until we find a package.json that contains the "name" and "version" fields.
+   */
+  private findExternalPackageJson(
+    packageName: string,
+    relativeToDir: string
+  ): PackageJson | null {
+    // The package.json is directly resolvable
+    const packageJsonPath = resolveRelativeToDir(
+      join(packageName, 'package.json'),
+      relativeToDir
+    );
+    if (packageJsonPath) {
+      return readJsonFile(packageJsonPath);
+    }
+
+    try {
+      // Resolve the main entry point of the package
+      const mainPath = resolveRelativeToDir(packageName, relativeToDir);
+      let dir = dirname(mainPath);
+
+      while (dir !== parse(dir).root) {
+        const packageJsonPath = join(dir, 'package.json');
+        if (existsSync(packageJsonPath)) {
+          const parsedPackageJson = readJsonFile(packageJsonPath);
```

Comment:
It's better to try and fail which will only open up the fs once

```
try {
  const parsedPackageJson = readJsonFile(path);
  if (parsedPackageJson.name && parsedPackageJson.version) {
    return parsedPackageJson
  }
} catch {
  // Either it doesn't exist.. or.. it couldn't be parsed.. but the one from package managers should be able to be parsed so it's the same as it doesn't exist.
}
```

Reply:
Hmm, no, we want to keep traversing upwards through directories which don't have package.json files in them and this change would cause it to blow up on the first parent directory that didn't have one?

Real example from today inside `minimatch`:

```
minimatch/dist/cjs/package.json
```

Means we need to go through

```
minimatch/dist/cjs
minimatch/dist
```

before reaching `minimatch/package.json`